README.md
<a href="http://www.ffuenf.de" title="ffuenf - code • design • e-commerce"><img src="https://github.com/ffuenf/Ffuenf_Common/blob/master/skin/adminhtml/default/default/ffuenf/ffuenf.png" alt="ffuenf - code • design • e-commerce" /></a>

Ffuenf_UpdateScripts
====================
[![GitHub tag](https://img.shields.io/github/tag/ffuenf/Ffuenf_UpdateScripts.svg)][tag]
[![Build Status](https://img.shields.io/travis/ffuenf/Ffuenf_UpdateScripts.svg)][travis]
[![Code Quality](https://scrutinizer-ci.com/g/ffuenf/Ffuenf_UpdateScripts/badges/quality-score.png)][code_quality]
[![Code Coverage](https://scrutinizer-ci.com/g/ffuenf/Ffuenf_UpdateScripts/badges/coverage.png)][code_coverage]
[![PayPal Donate](https://img.shields.io/badge/paypal-donate-blue.svg)][paypal_donate]
[tag]: https://github.com/ffuenf/Ffuenf_UpdateScripts
[travis]: https://travis-ci.org/ffuenf/Ffuenf_UpdateScripts
[code_quality]: https://scrutinizer-ci.com/g/ffuenf/Ffuenf_UpdateScripts
[code_coverage]: https://scrutinizer-ci.com/g/ffuenf/Ffuenf_UpdateScripts
[paypal_donate]: https://www.paypal.com/cgi-bin/webscr?cmd=_s-xclick&hosted_button_id=J2PQS2WLT2Y8W&item_name=Magento%20Extension%3a%20Ffuenf_UpdateScripts&item_number=Ffuenf_UpdateScripts&currency_code=EUR

This is a extension for Magento Community Edition that is meant to be a boilerplate to use independent upgrade scripts.

Platform
--------

The following versions are supported and tested:

* Magento Community Edition 1.9.2.4
* Magento Community Edition 1.9.1.1
* Magento Community Edition 1.8.1.0
* Magento Community Edition 1.7.0.2
* Magento Community Edition 1.6.2.0

Other versions are assumed to work.

Requirements
------------

|                                                                     | PHP 5.5           | PHP 5.6           | PHP 7.0           |
| ------------------------------------------------------------------- | ----------------- | ----------------- | ----------------- |
| [EOL](https://secure.php.net/supported-versions.php) / STABLE / RC  | STABLE            | **STABLE**        | **STABLE**        |
| automated tests on [travis]                                         | **required pass** | **required pass** | **required pass** |

Magento Community Edition officially supports PHP 5.4 and PHP 5.5.

Non-official compatibility to PHP 5.6 may be reached by following the tips on [Use of iconv.internal_encoding is deprecated](https://magento.stackexchange.com/questions/34015/magento-1-9-php-5-6-use-of-iconv-internal-encoding-is-deprecated).
Non-official compatibility to PHP 7.0 may be reached by using [Inchoo_PHP7](https://github.com/Inchoo/Inchoo_PHP7).

Installation
------------

Use [modman](https://github.com/colinmollenhour/modman) to install:
```
modman init
modman clone https://github.com/ffuenf/Ffuenf_Common
modman clone https://github.com/ffuenf/Ffuenf_UpdateScripts
```

Deinstallation
--------------

#### via [Ffuenf_MageTrashApp](https://github.com/ffuenf/Ffuenf_MageTrashApp)

An additional module called [Ffuenf_MageTrashApp](https://github.com/ffuenf/Ffuenf_MageTrashApp) has been installed to help you to uninstall this extension in a clean way.
If it is not yet installed, please install it from [Ffuenf_MageTrashApp](https://github.com/ffuenf/Ffuenf_MageTrashApp)
If it is installed, go to your backend menu System > Configuration > Advanced > MageTrashApp, then click on the tab "Extension Installed", select the drop down option "Uninstall" of this extension and press "Save Config" button to uninstall
If you use this extension, you don't need to make any queries in your database as explained below in case of manually uninstallation.

#### via [modman](https://github.com/colinmollenhour/modman)

Use [modman](https://github.com/colinmollenhour/modman) to clear all files and symlinks:
```
modman clean Ffuenf_UpdateScripts
```
see `sql/ffuenf_updatescripts_setup/uninstall.sql` to clear all entries of this extension from your database.

Development
-----------
1. Fork the repository from GitHub.
2. Clone your fork to your local machine:

        $ git clone https://github.com/USER/Ffuenf_UpdateScripts

3. Create a git branch

        $ git checkout -b my_bug_fix

4. Make your changes/patches/fixes, committing appropriately
5. Push your changes to GitHub
6. Open a Pull Request

License and Author
------------------

- Author:: Achim Rosenhagen (<a.rosenhagen@ffuenf.de>)
- Copyright:: 2016, ffuenf

The MIT License (MIT)

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.

CHANGELOG.md
# CHANGELOG for Ffuenf_UpdateScripts

This file is used to list changes made in each version of Ffuenf_UpdateScripts.

## 1.0.4 (February 3, 2016)

* add helper for adding/updating cms pages, static blocks and email templates

## 1.0.3 (December 20, 2015)

* fix dependency to Ffuenf_Common
* phpcs / code style fixes
* update travis build-matrix

## 1.0.2 (November 22, 2015)

* add uninstallation via [Ffuenf_MageTrashApp](https://github.com/ffuenf/Ffuenf_MageTrashApp)

## 1.0.1 (November 19, 2015)

* add dependency to [Ffuenf_Common](https://github.com/ffuenf/Ffuenf_Common)
* common extension structure
* add more tests

## 1.0.0 (November 8, 2015)

* initial public release


app/code/local/Ffuenf/UpdateScripts/Test/Config/Main.php
<?php
/**
 * Ffuenf_UpdateScripts extension.
 *
 * NOTICE OF LICENSE
 *
 * This source file is subject to the MIT License
 * that is bundled with this package in the file LICENSE.txt.
 * It is also available through the world-wide-web at this URL:
 * http://opensource.org/licenses/mit-license.php
 *
 * @category   Ffuenf
 *
 * @author     Achim Rosenhagen <a.rosenhagen@ffuenf.de>
 * @copyright  Copyright (c) 2015 ffuenf (http://www.ffuenf.de)
 * @license    http://opensource.org/licenses/mit-license.php MIT License
 */

class Ffuenf_UpdateScripts_Test_Config_Main extends EcomDev_PHPUnit_Test_Case_Config
{
    /**
     * Check if the installed module has the correct module version
     *
     * @test
     */
    public function testModuleConfig()
    {
        $this->assertModuleVersionGreaterThanOrEquals($this->expected('module')->getVersion(), 'module is new enough');
        $this->assertModuleCodePool($this->expected('module')->getCodePool(), 'correct module code pool');
        $this->assertModuleIsActive('module is active');
    }

    /**
     * Tests whether extension helper aliases are returning the correct class names
     *
     * @test
     */
    public function testHelperAliases()
    {
        $this->assertHelperAlias(
            'ffuenf_updatescripts',
            'Ffuenf_UpdateScripts_Helper_Data'
        );
    }

    /**
     * Tests whether extension uses the old-style admin routing (not compatible with SUPEE-6788).
     *
     * @test
     */
    public function testGetOldAdminRouting()
    {
        $routers = Mage::getConfig()->getNode('admin/routers');
        $offendingExtensions = array();
        foreach ($routers[0] as $router) {
            $name = $router->args->module;
            if ($name != 'Mage_Adminhtml') {
                $offendingExtensions[] = $router->args->module;
            }
        }
        $this->assertEquals(
            count($offendingExtensions),
            0,
            'This extension uses old-style admin routing which is not compatible with SUPEE-6788 / Magento 1.9.2.2+'
        );
    }
}


app/code/local/Ffuenf/UpdateScripts/Test/Helper/Data.php
<?php
/**
 * Ffuenf_UpdateScripts extension.
 *
 * NOTICE OF LICENSE
 *
 * This source file is subject to the MIT License
 * that is bundled with this package in the file LICENSE.txt.
 * It is also available through the world-wide-web at this URL:
 * http://opensource.org/licenses/mit-license.php
 *
 * @category   Ffuenf
 *
 * @author     Achim Rosenhagen <a.rosenhagen@ffuenf.de>
 * @copyright  Copyright (c) 2015 ffuenf (http://www.ffuenf.de)
 * @license    http://opensource.org/licenses/mit-license.php MIT License
 */

/**
 * @see Ffuenf_UpdateScripts_Helper_Data
 *
 */

class Ffuenf_UpdateScripts_Test_Helper_Data extends EcomDev_PHPUnit_Test_Case
{
    /**
     * Tests whether extension uses the old-style admin routing (not compatible with SUPEE-6788).
     *
     * @test
     */
    public function testGetOldAdminRouting()
    {
        $routers = Mage::getConfig()->getNode('admin/routers');
        $offendingExtensions = array();
        foreach ($routers[0] as $router) {
            $name = $router->args->module;
            if ($name != 'Mage_Adminhtml') {
                $offendingExtensions[] = $router->args->module;
            }
        }
        $this->assertEquals(
            count($offendingExtensions),
            0,
            'This extension uses old-style admin routing which is not compatible with SUPEE-6788 / Magento 1.9.2.2+'
        );
    }
}


app/code/local/Ffuenf/UpdateScripts/Test/Config/Setup.php
<?php
/**
 * Ffuenf_UpdateScripts extension.
 *
 * NOTICE OF LICENSE
 *
 * This source file is subject to the MIT License
 * that is bundled with this package in the file LICENSE.txt.
 * It is also available through the world-wide-web at this URL:
 * http://opensource.org/licenses/mit-license.php
 *
 * @category   Ffuenf
 *
 * @author     Achim Rosenhagen <a.rosenhagen@ffuenf.de>
 * @copyright  Copyright (c) 2015 ffuenf (http://www.ffuenf.de)
 * @license    http://opensource.org/licenses/mit-license.php MIT License
 */

class Ffuenf_UpdateScripts_Test_Config_Setup extends EcomDev_PHPUnit_Test_Case_Config
{
    /**
     * Check if setup resources are defined
     *
     * @test
     */
    public function testSetupDefined()
    {
        $this->assertSetupResourceDefined();
        $this->assertSchemeSetupExists();
    }

    /**
     * Check if update scripts exists for the correct module version
     *
     * @test
     */
    public function testSetupExists()
    {
        $this->assertSchemeSetupScriptVersions(
            '1.0.0', $this->expected('module')->getVersion(), null, 'ffuenf_updatescripts_setup'
        );
    }
}

app/code/local/Ffuenf/UpdateScripts/Helper/Data.php
<?php
/**
 * Ffuenf_UpdateScripts extension.
 *
 * NOTICE OF LICENSE
 *
 * This source file is subject to the MIT License
 * that is bundled with this package in the file LICENSE.txt.
 * It is also available through the world-wide-web at this URL:
 * http://opensource.org/licenses/mit-license.php
 *
 * @category   Ffuenf
 *
 * @author     Achim Rosenhagen <a.rosenhagen@ffuenf.de>
 * @copyright  Copyright (c) 2015 ffuenf (http://www.ffuenf.de)
 * @license    http://opensource.org/licenses/mit-license.php MIT License
 */

class Ffuenf_UpdateScripts_Helper_Data extends Ffuenf_Common_Helper_Core
{
    const CONFIG_EXTENSION_ACTIVE = 'ffuenf_updatescripts/general/enabled';

    /**
     * @var array
     */
    protected $_localeTemplatePath = array();

    /**
     * Variable for if the extension is active.
     *
     * @var bool
     */
    protected $_bExtensionActive;

    /**
     * Check to see if the extension is active.
     *
     * @return bool
     */
    public function isExtensionActive()
    {
        return $this->getStoreFlag(self::CONFIG_EXTENSION_ACTIVE, '_bExtensionActive');
    }

    /**
     * Takes a serialized data and do a search/replace in it
     *
     * @param string $search_for    string to search for (which should be replaced)
     * @param string $replace_with  string to replace search for with
     * @param string $data          serialized data to search in
     *
     * @return string updated serialized data
     */
    public function updateSerializedData($search_for, $replace_with, $data)
    {
        $unserialized = unserialize($data);
        recursive_array_replace($search_for, $replace_with, $unserialized);
        $edited_serialized_data = serialize($unserialized);
        return $edited_serialized_data;
    }

    /**
     * Takes an array and add it (merges it) with a serialized data
     *
     * @param array $add_to_array  array which should be included in the serialized data
     * @param string $data         serialized data to update
     *
     * @return string updated serialized data
     */
    public function addToSerializedData($add_to_array, $data)
    {
        $unserialized = unserialize($data);
        $merged = array_merge($unserialized, $add_to_array);
        $edited_serialized_data = serialize($merged);
        return $edited_serialized_data;
    }

    /**
     * Takes a serialized data and removes an entry in it
     *
     * @param string $search_for_array    string to search for (which should be replaced)
     * @param string $data          serialized data to search in
     *
     * @return string updated serialized data
     */
    public function removeFromSerializedData($search_for_array, $data)
    {
        $unserialized = unserialize($data);
        foreach ($search_for_array as $search_for) {
            $this->recursive_array_remove($search_for, $unserialized);
        }
        $edited_serialized_data = serialize($unserialized);
        return $edited_serialized_data;
    }

    /**
     * Helper function for updateSerializedData
     *
     * @param string $find     string to search for (which should be replaced)
     * @param string $replace  string to replace search for with
     * @param string &$data    serialized data to search in
     *
     * @return void
     */
    protected function recursive_array_replace($find, $replace, &$data) {
        if (is_array($data)) {
            foreach ($data as $key => $value) {
                if (is_array($value)) {
                    recursive_array_replace($find, $replace, $data[$key]);
                } else {
                    // have to check if it's string to ensure no switching to string for booleans/numbers/nulls - don't need any nasty conversions
                    if (is_string($value)) {
                        $data[$key] = str_replace($find, $replace, $value);
                    }
                }
            }
        } else {
            if (is_string($data)) {
                $data = str_replace($find, $replace, $data);
            }
        }
    }

    /**
     * Helper function for removeFromSerializedData
     *
     * @param string &$data serialized data to search in
     *
     * @return void
     */
    protected function recursive_array_remove($search_for, &$data) {
        foreach ($data as $key => $value) {
            if (is_array($value)) {
                $this->recursive_array_remove($search_for, $data[$key]);
            } else {
                if (strpos($key, $search_for) !== false) {
                    unset($data[$key]);
                }
            }
        }
    }

    /**
     * Retrieve template path for given locale
     *
     * @param  string $locale Locale
     * @param  string $type
     * @return string Locale Template Path
     */
    public function getLocaleTemplatePath($locale = 'de_DE', $type = 'block')
    {
        if (!isset($this->_localeTemplatePath[$locale][$type])) {
            $_localeTemplatePath = 'app' . DS . 'locale' . DS . $locale . DS . 'template' . DS . $type . DS;
            $this->_localeTemplatePath[$locale][$type] = Mage::getBaseDir() . DS . $_localeTemplatePath;
            if (!is_dir($this->_localeTemplatePath[$locale][$type])) {
                throw new Ffuenf_Common_Exception('Directory ' . $this->_localeTemplatePath[$locale][$type] . ' not found. Locale not installed?');
            }
        }
        return $this->_localeTemplatePath[$locale][$type];
    }

    /**
     * Get template content
     *
     * @param  string $identifier
     * @param  string $locale Locale
     * @param  string $type
     * @return string Template content
     */
    public function getTemplateContent($identifier, $locale = 'de_DE', $type = 'block')
    {
        $filepath = $this->getLocaleTemplatePath($locale, $type);
        $filename = $filepath . DS . $identifier . '.html';
        return @file_get_contents($filename);
    }

    /**
     * add cms pages, static blocks or transactional email templates
     *
     * @param  array $data
     * @param  int $storeId store-id
     * @param  string $type whether its a page, static block or transactional email template
     */
    public function saveCmsData($data, $storeId, $type = 'block') {
        $store = array($storeId);
        $store[] = Mage_Core_Model_App::ADMIN_STORE_ID;
        switch ($type) {
            case 'email':
                $template = Mage::getModel('core/email_template')->loadByCode($data['identifier']);
                if (preg_match('/<!--@subject\s*(.*?)\s*@-->/u', $data['content'], $matches)) {
                    $template->setTemplateSubject($matches[1]);
                    $data['content'] = str_replace($matches[0], '', $data['content']);
                }
                if (preg_match('/<!--@vars\s*((?:.)*?)\s*@-->/us', $data['content'], $matches)) {
                    $data['content'] = str_replace($matches[0], '', $data['content']);
                }
                if (preg_match('/<!--@styles\s*(.*?)\s*@-->/s', $data['content'], $matches)) {
                    $template->setTemplateStyles($matches[1]);
                    $data['content'] = str_replace($matches[0], '', $data['content']);
                }
                $data['content'] = preg_replace('#\{\*.*\*\}#suU', '', $data['content']);
                $template->setTemplateCode($data['identifier'])
                            ->setTemplateType(2)
                            ->setAddedAt(Mage::getSingleton('core/date')->gmtDate())
                            ->setModifiedAt(Mage::getSingleton('core/date')->gmtDate());
                $template->setTemplateText($data['content'])->save();
                Ffuenf_Common_Model_Logger::logSystem(
                    array(
                        'class'   => 'Ffuenf_UpdateScripts',
                        'level'   => Zend_Log::DEBUG,
                        'message' => 'Created new E-Mail Template: ' . $data['title'],
                        'details' => $data['content']
                    )
                );
                break;
            case 'page':
            case 'block':
            default:
                $model = ($type == 'page') ? Mage::getModel('cms/page') : Mage::getModel('cms/block');
                if (preg_match('/<!--@title\s*(.*?)\s*@-->/u', $data['content'], $matches)) {
                    $data['title'] = $matches[1];
                    $data['content'] = str_replace($matches[0], '', $data['content']);
                }
                if (preg_match('/<!--@heading\s*(.*?)\s*@-->/u', $data['content'], $matches)) {
                    $data['content_heading'] = $matches[1];
                    $data['content'] = str_replace($matches[0], '', $data['content']);
                }
                if (preg_match('/<!--@root_template\s*(.*?)\s*@-->/s', $data['content'], $matches)) {
                    $data['root_template'] = $matches[1];
                    $data['content'] = str_replace($matches[0], '', $data['content']);
                }
                if (preg_match('/<!--@view_mode\s*(.*?)\s*@-->/s', $data['content'], $matches)) {
                    $data['view_mode'] = $matches[1];
                    $data['content'] = str_replace($matches[0], '', $data['content']);
                }
                $data['content'] = preg_replace('#\{\*.*\*\}#suU', '', $data['content']);
                $collection = $model->getCollection()
                                    ->addFieldToFilter('identifier', $data['identifier'])
                                    ->addStoreFilter($storeId)
                                    ->addFieldToFilter('store_id', array('in' => $store));
                $cmsItem = $collection->getFirstItem();
                if ($cmsItem && ($cmsItem->getBlockId() || $cmsItem->getPageId())) {
                    $cmsItem->load();
                    $oldData = $cmsItem->getData();
                    $data = array_merge($oldData, $data);
                    $cmsItem->setData($data)->save();
                } else {
                    $model->setData($data)->save();
                }
                Ffuenf_Common_Model_Logger::logSystem(
                    array(
                        'class'   => 'Ffuenf_UpdateScripts',
                        'level'   => Zend_Log::DEBUG,
                        'message' => 'Created new CMS ' . $type . ': ' . $data['title'],
                        'details' => $data['content']
                    )
                );
                break;
        }
        return $this;
    }
}


app/code/local/Ffuenf/UpdateScripts/sql/ffuenf_updatescripts_setup/install-1.0.0.php
<?php
/**
 * Ffuenf_UpdateScripts extension.
 *
 * NOTICE OF LICENSE
 *
 * This source file is subject to the MIT License
 * that is bundled with this package in the file LICENSE.txt.
 * It is also available through the world-wide-web at this URL:
 * http://opensource.org/licenses/mit-license.php
 *
 * @category   Ffuenf
 *
 * @author     Achim Rosenhagen <a.rosenhagen@ffuenf.de>
 * @copyright  Copyright (c) 2015 ffuenf (http://www.ffuenf.de)
 * @license    http://opensource.org/licenses/mit-license.php MIT License
 */

/*
 * for reference see
 * http://www.webguys.de/magento/eav-attribute-setup/
 * http://www.webguys.de/magento/adventskalender/turchen-23-installscripte-in-magento/
 */
/* @var $installer Mage_Core_Model_Resource_Setup */
$installer = $this;
$conn = $installer->getConnection();

$installer->startSetup();

$installer->endSetup();

app/code/local/Ffuenf/UpdateScripts/data/ffuenf_updatescripts_setup/data-install-1.0.0.php
<?php
/*
 * for reference see
 * http://www.webguys.de/magento/eav-attribute-setup/
 * http://www.webguys.de/magento/adventskalender/turchen-23-installscripte-in-magento/
 */
$installer = $this;
$conn = $installer->getConnection();

$installer->startSetup();

/*
 * compatibility to SUPEE-6788 / Magento 1.9.2.2+
 *
$installer->setConfigData('admin/security/extensions_compatibility_mode', '0', 'default', '0');
$blockNames = array(
    ''
);
foreach ($blockNames as $blockName) {
    $whitelistBlock = Mage::getModel('admin/block')->load($blockName, 'block_name');
    $whitelistBlock->setData('block_name', $blockName);
    $whitelistBlock->setData('is_allowed', 1);
    $whitelistBlock->save();
}
$variableNames = array(
    ''
);
foreach ($variableNames as $variableName) {
    $whitelistVar = Mage::getModel('admin/variable')->load($variableName, 'variable_name');
    $whitelistVar->setData('variable_name', $variableName);
    $whitelistVar->setData('is_allowed', 1);
    $whitelistVar->save();
}
*/

/*
 * add transactional email template
 *
$identifier      = '_identifier';
$storeId         = 1;
$locale          = 'en_US';
$templateContent = Mage::helper('ffuenf_updatescripts')->getTemplateContent($identifier, $locale, 'email');
$cmsPage         = array(
                    'title'      => 'Your title',
                    'identifier' => $identifier,
                    'content'    => $templateContent
                 );
Mage::helper('ffuenf_updatescripts')->saveCmsData($cmsPage, $storeId, 'email');
*/

/*
 * add cms static block
 *
$identifier      = '_identifier';
$storeId         = 1;
$locale          = 'en_US';
$templateContent = Mage::helper('ffuenf_updatescripts')->getTemplateContent($identifier, $locale, 'block');
$cmsBlock        = array(
                    'title'      => 'Your title',
                    'identifier' => $identifier,
                    'content'    => $templateContent,
                    'is_active'  => 1,
                    'sort_order' => 0,
                    'stores'     => array($storeId),
                    'view_mode'  => 'core' # or 'easytemplate' for Webguys_EasyTemplate
                 );
Mage::helper('ffuenf_updatescripts')->saveCmsData($cmsBlock, $storeId, 'block');
*/

/*
 * add cms page
 *
$identifier      = '_identifier';
$storeId         = 1;
$locale          = 'en_US';
$templateContent = Mage::helper('ffuenf_updatescripts')->getTemplateContent($identifier, $locale, 'page');
$cmsPage         = array(
                    'title'      => 'Your title',
                    'identifier' => $identifier,
                    'content'    => $templateContent,
                    'is_active'  => 1,
                    'sort_order' => 0,
                    'stores'     => array($storeId),
                    'view_mode'  => 'core' # or 'easytemplate' for Webguys_EasyTemplate
                 );
Mage::helper('ffuenf_updatescripts')->saveCmsData($cmsPage, $storeId, 'page');
*/

$installer->endSetup();